codegenerator/cli/templates/static/codegen/src/globalState/GlobalState.res

```diff
@@ -790,66 +775,83 @@ let invalidatedActionReducer = (state: t, action: action) =>
     (state, [])
   }
 
-let executePartitionQuery = (
-  query,
+let executeQuery = (
+  query: FetchState.query,
   ~logger,
   ~chainWorker,
   ~currentBlockHeight,
   ~chain,
-  ~dispatchAction,
   ~isPreRegisteringDynamicContracts,
 ) => {
-  let module(ChainWorker: ChainWorker.S) = chainWorker
-
-  let logger = Logging.createChildFrom(
-    ~logger,
-    ~params={
-      "chainId": chain->ChainMap.Chain.toChainId,
-      "logType": "Block Range Query",
-      "workerType": ChainWorker.name,
-    },
-  )
-  let logger = query->FetchState.getQueryLogger(~logger)
-  ChainWorker.fetchBlockRange(
-    ~query,
-    ~logger,
-    ~currentBlockHeight,
-    ~isPreRegisteringDynamicContracts,
-  )->Promise.thenResolve(res =>
-    switch res {
-    | Ok(res) => dispatchAction(BlockRangeResponse(chain, res))
-    | Error(e) => dispatchAction(ErrorExit(e))
-    }
-  )
+  switch query {
+  | PartitionQuery({partitionId, fromBlock, toBlock, contractAddressMapping}) =>
+    chainWorker->ChainWorker.fetchBlockRange(
+      ~fromBlock,
+      ~toBlock,
+      ~contractAddressMapping,
+      ~partitionId,
+      ~chain,
+      ~currentBlockHeight,
+      ~isPreRegisteringDynamicContracts,
+      ~logger,
+      //Only apply wildcards on the first partition
+      //to avoid duplicate wildcard queries
+      ~shouldApplyWildcards=partitionId === "0",
+    )
+  | MergeQuery({partitionId, fromBlock, toBlock, contractAddressMapping}) =>
+    chainWorker->ChainWorker.fetchBlockRange(
+      ~fromBlock,
+      ~toBlock=Some(toBlock),
+      ~contractAddressMapping,
+      ~partitionId,
+      ~chain,
+      ~currentBlockHeight,
+      ~isPreRegisteringDynamicContracts,
+      ~logger,
+      //Only apply wildcards on the first partition
+      //to avoid duplicate wildcard queries
+      ~shouldApplyWildcards=partitionId === "0",
+    )
+  }
 }
 
 let checkAndFetchForChain = (
   //Used for dependency injection for tests
   ~waitForNewBlock,
-  ~executePartitionQuery,
+  ~executeQuery,
   //required args
   ~state,
   ~dispatchAction,
 ) => async chain => {
   let chainFetcher = state.chainManager.chainFetchers->ChainMap.get(chain)
   if !isRollingBack(state) {
-    let {chainConfig: {chainWorker}, logger, currentBlockHeight, partitionedFetchState} = chainFetcher
-
-    await chainFetcher.sourceManager->SourceManager.fetchBatch(
-      ~allPartitions=partitionedFetchState.partitions,
-      ~waitForNewBlock=(~currentBlockHeight, ~logger) => chainWorker->waitForNewBlock(~currentBlockHeight, ~logger),
-      ~onNewBlock=(~currentBlockHeight) => dispatchAction(FinishWaitingForNewBlock({chain, currentBlockHeight})),
+    let {
+      chainConfig: {chainWorker},
+      logger,
+      currentBlockHeight,
+      partitionedFetchState,
+    } = chainFetcher
+
+    await chainFetcher.sourceManager->SourceManager.fetchNext(
+      ~fetchState=partitionedFetchState.partitions->Js.Array2.unsafe_get(0),
```

Comment:
I think lets get rid of partitionedFetchState ASAP

Reply:
Let's do it in the next PR